### PR TITLE
Bug fixes and error handling in Redis - Vectorstore

### LINKED ISSUE #3893: AttributeError: 'Redis' object has no attribute 'module_list'
I'm trying to make a vectorstore using redis and store the embeddings in redis.
When I write the code
rds = Redis.from_documents(docs, embeddings, redis_url="redis://localhost:6379",  index_name='test_link')
I get the following error
AttributeError: 'Redis' object has no attribute 'module_list'.
Note: I'm trying to run redis locally on windows subsystem ubuntu.
Please help.

### PR DESCRIPTION
**Pull request by 0xcaffeinated**
# Bug fixes in Redis - Vectorstore (Added the version of redis to the error message and removed the cls argument from a classmethod)

Fixes  #3893 #4896

## Who can review?

@dev2049 @tylerhutcherson

### CODE REVIEW

File: langchain/vectorstores/redis.py [hunk 1]
```diff
@@ -33,36 +33,10 @@
...
-REDIS_REQUIRED_MODULES = [
-    {"name": "search", "ver": 20400},
-    {"name": "searchlight", "ver": 20400},
-]
-
 # distance mmetrics
 REDIS_DISTANCE_METRICS = Literal["COSINE", "IP", "L2"]
 
 
-def _check_redis_module_exist(client: RedisType, required_modules: List[dict]) -> None:
```
Comment (tylerhutcherson) [inline langchain/vectorstores/redis.py:46]: This is not something we should remove. This checks for the existence of the required Redis extensions on the database side (RediSearch: https://redis.io/docs/stack/search).

**Review by tylerhutcherson — CHANGES_REQUESTED** [review]
Hey -- What is this fixing exactly? We need this redis module check here otherwise it will fail for folks running vanilla redis

**Comment by tylerhutcherson** [comment]
> Hey -- What is this fixing exactly? We need this redis module check here otherwise it will fail for folks running vanilla redis

The confusion might be that there's Redis modules... RediSearch and RedisJSON for example (part of Redis Stack). These are what we use the code above to check for existence...

And then there's Python client libraries. The standard `redis` python library is all we need for langchain.

**Comment by iamadhee** [comment]
I have reverted the previous code deletion and refined the error message @tylerhutcherson

**Comment by iamadhee** [comment]
@nikhilkandur Please do `pip uninstall redisearch` and  do `pip install redis==4.5.5`
that'll solve the issue.

**Comment by tylerhutcherson** [comment]
> Sorry about all the chaos. I'll revert the changes done to this error message. I'll just probably add the version of redis (>=4.1.0) that has to be installed by the user, in order to be compatible with langchain [here](https://github.com/hwchase17/langchain/blob/1ed4228822498fd58d33d377572af0821c496b20/langchain/vectorstores/redis.py#L129). Would that be fine?

Perfect. I think that will help a ton.

**Comment by nikhilkandur** [comment]
I installed `redis==4.5.5` but i'm getting below error but that version of [redisearch](https://pypi.org/project/redisearch/#history) doesn't exists.

```sh
ValueError: Redis failed to connect: You must add the RediSearch (>= 2.4) module from Redis Stack. Please refer to Redis Stack docs: https://redis.io/docs/stack/
```

**Comment by tylerhutcherson** [comment]
> @nikhilkandur Here's what:
> 
> The [Redisearch](https://pypi.org/project/redisearch/#history) standalone project is discontinued. We won't need the`redisearch` python package to be installed as it is already available bundled within the standard `redis>=4.1.0` package.
> 
> So, the error you get right now despite what the message suggests has nothing to do with redisearch (you can refer the previous conversations as well). I doubt it might have something to do with the redis connection itself. Please check on that front.


Exactly. The error you have @nikhilkandur is because you need to run the [Redis Stack docker container](https://redis.io/docs/stack/get-started/install/docker/) (or sign up for Redis Cloud) that bundles RediSearch functionality into the database. The Python client side is different. `redis>=4.1.0` is sufficient.

**Comment by nikhilkandur** [comment]
> > @nikhilkandur Here's what:
> > The [Redisearch](https://pypi.org/project/redisearch/#history) standalone project is discontinued. We won't need the`redisearch` python package to be installed as it is already available bundled within the standard `redis>=4.1.0` package.
> > So, the error you get right now despite what the message suggests has nothing to do with redisearch (you can refer the previous conversations as well). I doubt it might have something to do with the redis connection itself. Please check on that front.
> 
> Exactly. The error you have @nikhilkandur is because you need to run the [Redis Stack docker container](https://redis.io/docs/stack/get-started/install/docker/) (or sign up for Redis Cloud) that bundles RediSearch functionality into the database. The Python client side is different. `redis>=4.1.0` is sufficient.

Currently, I'm making use of redis addon for heroku and using redis>=4.1.0 at the client side to make the vector store work. Isn't it sufficient ?

**Comment by tylerhutcherson** [comment]
> > > @nikhilkandur Here's what:
> > > The [Redisearch](https://pypi.org/project/redisearch/#history) standalone project is discontinued. We won't need the`redisearch` python package to be installed as it is already available bundled within the standard `redis>=4.1.0` package.
> > > So, the error you get right now despite what the message suggests has nothing to do with redisearch (you can refer the previous conversations as well). I doubt it might have something to do with the redis connection itself. Please check on that front.
> > 
> > 
> > Exactly. The error you have @nikhilkandur is because you need to run the [Redis Stack docker container](https://redis.io/docs/stack/get-started/install/docker/) (or sign up for Redis Cloud) that bundles RediSearch functionality into the database. The Python client side is different. `redis>=4.1.0` is sufficient.
> 
> Currently, I'm making use of redis addon for heroku and using redis>=4.1.0 at the client side to make the vector store work. Isn't it sufficient ?

Sounds like Heroku's redis implementation does not include any of the add-on modules like RediSearch, RedisJSON, RedisGraph, etc. But to use Redis as a vector database, you need the RediSearch module...

**Comment by nikhilkandur** [comment]
There seems like another bug, not sure if it is fixed recently in unreleased version. I'm seeing `kwargs` field inside kwargs. You can see it [here](https://github.com/hwchase17/langchain/blob/master/langchain/vectorstores/redis.py#L406) while instantiation

```sh
  File ".../site-packages/redis/connection.py", line 956, in __init__
    super().__init__(**kwargs)
TypeError: AbstractConnection.__init__() got an unexpected keyword argument 'kwargs'
```

**Comment by tylerhutcherson** [comment]
> There seems like another bug, not sure if it is fixed recently in unreleased version. I'm seeing `kwargs` field inside kwargs. You can see it [here](https://github.com/hwchase17/langchain/blob/master/langchain/vectorstores/redis.py#L406) while instantiation
> 
> ```shell
>   File ".../site-packages/redis/connection.py", line 956, in __init__
>     super().__init__(**kwargs)
> TypeError: AbstractConnection.__init__() got an unexpected keyword argument 'kwargs'
> ```

yup fixing this

**Comment by tylerhutcherson** [comment]
@iamadhee Can you change the name of this PR to be more specific to this fix and then will get this prioritized as well. Thanks!

**Comment by iamadhee** [comment]
> @iamadhee Can you change the name of this PR to be more specific to this fix and then will get this prioritized as well. Thanks!

Have added the new title and description

**Comment by iamadhee** [comment]
@tylerhutcherson updated the redisearch error message to avoid further confusions

**Comment by iamadhee** [comment]
Is there anything else that is needed from my end on this PR? @tylerhutcherson

File: langchain/vectorstores/redis.py [hunk 2]
```diff
@@ -56,8 +56,9 @@ def _check_redis_module_exist(client: RedisType, required_modules: List[dict]) -
             return
     # otherwise raise error
     error_message = (
-        "You must add the RediSearch (>= 2.4) module from Redis Stack. "
-        "Please refer to Redis Stack docs: https://redis.io/docs/stack/"
+        "Redis cannot be used as a vector database without Redisearch"
+        "Please head to https://redis.io/docs/stack/search/quick_start/"
+        "to know more about installing redisearch"
```
Comment (tylerhutcherson) [inline langchain/vectorstores/redis.py:61]:
```suggestion
        "to know more about installing the RediSearch module within Redis Stack."
```

File: langchain/vectorstores/redis.py [hunk 3]
```diff
@@ -56,8 +56,9 @@ def _check_redis_module_exist(client: RedisType, required_modules: List[dict]) -
             return
     # otherwise raise error
     error_message = (
-        "You must add the RediSearch (>= 2.4) module from Redis Stack. "
-        "Please refer to Redis Stack docs: https://redis.io/docs/stack/"
+        "Redis cannot be used as a vector database without Redisearch"
```
Comment (tylerhutcherson) [inline langchain/vectorstores/redis.py:59]:
```suggestion
        "Redis cannot be used as a vector database without RediSearch >=2.4"
```

**Review by tylerhutcherson — APPROVED** [review]
2 quick suggestions on wording. but yeah, LGTM!

cc @dev2049

**Comment by iamadhee** [comment]
Made the suggested changes cc @tylerhutcherson @dev2049

**Comment by dev2049** [comment]
thanks @iamadhee!

**Pull request merged**